### PR TITLE
Add validation for metrics config during parquet write

### PR DESCRIPTION
**Pull request by holdenk**
Validate that the metrics config refers to actual columns in the schema during write.

This is for issue #1548.

The discussion in the ticket suggested calling it during table creation, but I moved it to parquet write since we don't parse the metrics config until the write stage not in the create call. Still new to the code base so if that isn't the tradeoff we want hella ok to refactor.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 1]
```diff
@@ -70,10 +71,17 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
...
         });
-
     return spec;
   }
 
+  public void validateProperties(Schema schema) {
+    for (String column : columnModes.keySet()) {
+      if (schema.findField(column) == null) {
+        throw new ValidationException("Could not find column %s from metrics in schema %s", column, schema);
+      }
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:81]: You can use `ValidationException.check` like a `Precondition` to avoid adding lots of `if` statements:

```java
ValidationException.check(schema.findField(column) != null, "Could not find ...", ...);
```

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 2]
```diff
@@ -70,10 +71,17 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
           }
           spec.columnModes.put(columnAlias, mode);
         });
-
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:73]: Nit: this is an unnecessary whitespace change that could lead to git conflicts. Could you add it back?

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 3]
```diff
@@ -226,6 +226,9 @@ private CompressionCodecName codec() {
       set("parquet.avro.write-old-list-structure", "false");
       MessageType type = ParquetSchemaUtil.convert(schema, name);
 
+      // Check that our metrics make sense
+      metricsConfig.validateProperties(schema);
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:230]: Is this the right place to do the validation?

If a user adds a bad property or performs some schema update that causes a validation error, that would break all writes to the table. To me, it doesn't seem like we are catching the problem early enough and possibly allowing a typo to break scheduled jobs.

What do you think about adding this validation when altering the table? `UpdateProperties` could check whether any properties starting with `write.metadata.metrics` were modified and run this. Similarly, `UpdateSchema` could run this as well, although I think that we should probably modify `UpdateSchema` to simply update the properties for column renames (if that's easily done).

**Comment by rdblue** [comment]
Thanks for working on this, @holdenk! It would be great to get this gap fixed.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 1 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:81]: Oh nice :)

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 3 above)
Comment (holdenk) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:230]: Makes sense, I'll move this over to UpdateProperties/UpdateSchema tomorrow :)

**Comment by holdenk** [comment]
Thanks for the quick review :) I'll circle back tomorrow with the changes :)

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 3 above)
Comment (aokolnychyi) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:230]: I'd also cover table creation. I feel that we should fail table creation if the metrics config refers to non-existing columns. I have seen so many cases when people clone an example but forget to update the metrics config and they realize the problem only after ingestion is done.

Comment (rdblue) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:230]: That's a good point. Table creation is a good time to validate this as well.

Comment (holdenk) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: I've put this in for table creation & schema + property evolution. Let me know if I hooked in the correct places.

**Comment by holdenk** [comment]
I think I've addressed the comments, although it's possible the hooks I've chosen are not the best places but it seems to cover the right areas with the tests. Let me know if there is a better place to put the hooks for validation.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java [hunk 4]
```diff
@@ -303,6 +304,9 @@ public Table create() {
       }
 
       Map<String, String> properties = propertiesBuilder.build();
+      // Validate the metrics
+      MetricsConfig.fromProperties(properties).validateProperties(schema);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: I think we need this in the `createTransaction` and other methods as well.

Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: Oh cool, I didn't think of putting it there, I'll give that a shot.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 5]
```diff
@@ -74,6 +75,14 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
     return spec;
   }
 
+  public void validateProperties(Schema schema) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:78]: I am not sure `validateProperties` is the most descriptive name in this case.

Something like `validateReferencedColumns` or similar would probably convey the purpose more. For example, we don't validate the property values here (if a value is invalid, it will be defaulted). We just check the config refers to valid columns.

Alternatively to changing the name, we can make this method static and accept both properties as well as schema. For example, we may invoke it as `MetricsConfig.validateProperties(props, schema)` and it will check both the column names as well as metrics modes (we can overload `fromProperties` so that it will throw an exception instead of defaulting).

Thoughts, @holdenk @rdblue @RussellSpitzer @shardulm94?

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:78]: I'd probably go for a static method and validating the value modes as well.

File: core/src/main/java/org/apache/iceberg/PropertiesUpdate.java [hunk 6]
```diff
@@ -89,6 +89,11 @@ public UpdateProperties defaultFormat(FileFormat format) {
 
     newProperties.putAll(updates);
 
+    // Validate the metrics
+    if (base.schema() != null) {
+      MetricsConfig.fromProperties(newProperties).validateProperties(base.schema());
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:94]: If we go for a static method, this will also become shorter:

```
MetricsConfig.validateProperties(newProperties, base.schema());
```

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java (hunk 4 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: @rdblue, will moving the check to the constructor require us to also update the metrics config when the schema evolves? Otherwise, drop/rename operations will always fail if the metrics config refers to them?

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: I think it is a good thing to update the metrics config implicitly when the schema evolves, BTW.

**Comment by aokolnychyi** [comment]
@rdblue @shardulm94 @RussellSpitzer, not related to this PR but a general question: the metrics config relies on column names instead of column ids as it is controlled through table properties. I was wondering whether it is safe to do so.

I considered the following cases:
- While writing new data, it should be always OK to follow the name-based approach if the config is up-to-date.
- While importing data (assuming we have a correct name mapping), we will assign correct ids to columns, use the ids to find the current aliases in the schema, so this should work.
- When fixing (re-importing) metadata (assuming we have renamed a column through Iceberg and no name mapping), we will just use the column id in the file to find the current name, so this should work.

Overall, it seems safe to me. Any cases I missed?

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 5 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:78]: I'm not sure about having this be a static method since it would end up constructing an instance of the object anyways, but I don't have strong feelings here. I lean towards giving it a name like `validateReferencedColumns` personally.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java (hunk 4 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: I think maybe automatically updating the metrics config on schema evolution could be a good follow up ticket, WDYT @aokolnychyi ?
I think for this first part requiring that the metrics config is valid during schema evolution is a good first step.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: I agree that it is a follow-up. This will catch the problem and we can rewrite later.

We may also want to consider alternative storage for these that uses ID rather than name. I guess we could rewrite to ID-based.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: It is absolutely a follow-up item. Just wanted to bring attention to that.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 5 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:78]: Same for me. I don't feel strongly about this one.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java (hunk 4 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: Sounds good. I've added https://github.com/apache/iceberg/issues/2079

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: Thanks, @holdenk!

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 5 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:78]: I think it is fine either way. I think `MetricsConfig#validateProperties` is reasonable, but I like `validateReferencedColumns` as well.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java (hunk 4 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:308]: After thinking about this a bit more, I don't think it is a good idea to add the check in the `TableMetadata` constructor after all. The problem with doing that is that it would break existing tables because they would validate when loading a table. Instead, we need to validate each case where a table is created so we could put it in the `newTableMetadata` factory method instead.

That would mean we also need to keep the validation in the property and schema update operations as you already have in this PR.

File: core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java [hunk 7]
```diff
@@ -155,6 +156,42 @@ public void testDeleteColumn() {
...
+        () ->
+          table.updateSchema()
+              .deleteColumn("id")
+          .commit());
+
+    String updatedJson = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    NameMapping updated = NameMappingParser.fromJson(updatedJson);
+
+    // should not change the mapping
+    validateUnchanged(mapping, updated);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:192]: It seems strange to me that this would be tested in the suite for the schema mapping. This should be independent of mapping. While it is okay to have a test that a mapping is not updated if the validation fails, that should be covered by not updating the schema if metrics property validation fails. Could you start a new suite similar to `TestSchemaAndMappingUpdate` for this instead of adding the tests here?

File: core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java [hunk 8]
```diff
@@ -155,6 +156,42 @@ public void testDeleteColumn() {
...
+          .commit());
+
+    AssertHelpers.assertThrows(
+        "Deleting a column with metrics fails",
+        ValidationException.class,
+        null,
+        () ->
+          table.updateSchema()
+              .deleteColumn("id")
+          .commit());
```
Comment (kbendick) [inline core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:186]: Nit: line indentation doesn't line up with the other values.

File: core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java (hunk 7 above)
Comment (holdenk) [reply core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:192]: Sure I can make a new suite for this.

Comment (holdenk) [reply core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:192]: Sure I'll add a new test suite next.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 9]
```diff
@@ -110,6 +110,11 @@ static TableMetadata newTableMetadata(Schema schema,
     int freshSortOrderId = sortOrder.isUnsorted() ? sortOrder.orderId() : INITIAL_SORT_ORDER_ID;
     SortOrder freshSortOrder = freshSortOrder(freshSortOrderId, freshSchema, sortOrder);
 
+    // Validate the metrics configuration. Note: we only do this on new tables to we don't
+    // break existing tables.
+    MetricsConfig.fromProperties(properties).validateReferencedColumns(schema);
+
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:117]: Nit: double newline.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java [hunk 10]
```diff
@@ -303,7 +303,8 @@ public Table create() {
       }
 
       Map<String, String> properties = propertiesBuilder.build();
-      TableMetadata metadata = tableMetadata(schema, spec, sortOrder, properties, location);
+
+      TableMetadata metadata = TableMetadata.newTableMetadata(schema, spec, sortOrder, location, properties);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:307]: Is it necessary to change this file? Doesn't `tableMetadata` call `newTableMetadata`?

File: spark/src/test/java/org/apache/iceberg/spark/source/TestWriteMetricsConfig.java [hunk 11]
```diff
@@ -228,6 +230,23 @@ public void testCustomMetricCollectionForParquet() throws IOException {
...
+    properties.put(TableProperties.DEFAULT_WRITE_METRICS_MODE, "counts");
+    properties.put("write.metadata.metrics.column.ids", "full");
+
+    AssertHelpers.assertThrows("Creating a table with invalid metrics should fail",
+        ValidationException.class,
+        null,
+        () -> tables.create(SIMPLE_SCHEMA, spec, properties, tableLocation));
+  }
+
+
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestWriteMetricsConfig.java:249]: Nit: extra newline.

File: core/src/main/java/org/apache/iceberg/PropertiesUpdate.java [hunk 12]
```diff
@@ -89,6 +89,11 @@ public UpdateProperties defaultFormat(FileFormat format) {
 
     newProperties.putAll(updates);
 
+    // Validate the metrics
+    if (base != null && base.schema() != null) {
+      MetricsConfig.fromProperties(newProperties).validateReferencedColumns(base.schema());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:94]: I don't think this is tested. Could you add a test to the property update tests?

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 13]
```diff
@@ -357,7 +357,14 @@ private void internalMove(String name, Move move) {
...
   public Schema apply() {
-    return applyChanges(schema, deletes, updates, adds, moves);
+    Schema newSchema = applyChanges(schema, deletes, updates, adds, moves);
+
+    // Validate the metrics if we have existing properties.
+    if (base != null && base.properties() != null) {
+      MetricsConfig.fromProperties(base.properties()).validateReferencedColumns(newSchema);
+    }
+
+    return newSchema;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:367]: I don't think that this path is tested. Could you add one?

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java [hunk 14]
```diff
@@ -74,6 +75,14 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
     return spec;
   }
 
+  public void validateReferencedColumns(Schema schema) {
+    for (String column : columnModes.keySet()) {
+      ValidationException.check(
+          schema.findField(column) != null,
+          "Could not find column %s from metrics in schema %s", column, schema);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetricsConfig.java:82]: Just thinking out loud: will be clear for an inexperienced user what is wrong?
Is `from metrics` descriptive enough? Should we make it `from metrics config`?
Or make it as `Invalid metrics config: could not find column %s in schema %s`?

**Comment by aokolnychyi** [comment]
LGTM apart from the existing comments.

File: core/src/main/java/org/apache/iceberg/MetricsConfig.java (hunk 14 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/MetricsConfig.java:82]: That seems like a fair criticism. What about adding the bad property name and noting that it is a table property? That would help.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java (hunk 10 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:307]: My bad, I'll revert this change.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java (hunk 13 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/SchemaUpdate.java:367]: It is currently covered by `testModificationWithMetricsMetrics`, commenting out the check results in the following error when running gradle build:

```
org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 1] FAILED
    java.lang.AssertionError: No exception was thrown (Deleting a column with metrics fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:179)

org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 2] FAILED
    java.lang.AssertionError: No exception was thrown (Deleting a column with metrics fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:179)

```

File: core/src/main/java/org/apache/iceberg/PropertiesUpdate.java (hunk 12 above)
Comment (holdenk) [reply core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:94]: This is covered in the same test as the other one (the exception is on a different line though):

```
org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 1] FAILED
    java.lang.AssertionError: No exception was thrown (Creating metrics for non-existent column fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:169)

org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 2] FAILED
    java.lang.AssertionError: No exception was thrown (Creating metrics for non-existent column fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:169)

```

**Comment by rdblue** [comment]
Thanks, @holdenk! Really nice to have this done.
